### DISCUSSION TITLE
[jest] spyOn return type now never for Storage prototype members

### DISCUSSION THREAD

**Discussion by wimbarelds** [DefinitelyTyped/DefinitelyTyped#63129] [Issues with a @types package] [answered]
Previously (`@types/jest@29.2.1`), in order to spy on `Storage.prototype.setItem`, we could:

```typescript
jest.spyOn(Storage.prototype, 'setItem').mockImplementation(jest.fn());
```

As of version `29.2.2` of `@types/jest`, this no longer works. The returntype of spyOn in this case is `never`.
This functionality does work in `jest@29.2.2` with `@types/jest@29.2.1`

[a comment by an automated account is not shown]

**Comment by mrazauskas** [comment]
Thanks for report! It can be that I introduced this issue in #62781. Will take a look.

**Comment by mrazauskas** [comment]
In the above mentioned PR, I copied copied `jest.Spied*` types from Jest repo. These are the typings which ship with Jest, to use them simply: `import {jest} from '@jest/globals'`.

Your example works as expected with types from Jest, but fails if I try to use definitions from `@types/jest`. Seems like there is nothing wrong with `jest.Spied*` types. To implement them I had to tweak `spyOn` overloads here in `@types/jest`. The overloads differ from the ones in Jest repo. The root of the issue are these overloads.

I try to fix them, but it may take time. If you are in a hurry just import from `@jest/globals` (;

**Reply by wimbarelds** [reply]
Thanks for looking into this! Looking at the PR you linked (which I had also looked at, I assumed the problem was to do with the order in which the overloads are organized, as that changed slightly. Though when I am reading the overload type definitions (which I admittedly have a tentative understanding of at best), that _seems_ like it shouldn't be the case.

**Comment by mrazauskas** [comment]
Looks like the index signature in `Storage` interface is causing this issue. Here is how the interface is declared:

```ts
interface Storage {
    readonly length: number;
    clear(): void;
    getItem(key: string): string | null;
    key(index: number): string | null;
    removeItem(key: string): void;
    setItem(key: string, value: string): void;
    [name: string]: any; // <- everything works as expected if this line gets removed
}

declare var Storage: {
    prototype: Storage;
    new(): Storage;
};
```

The implementation of index signature is slightly differs between `@types/jest` and build-in Jest’s types. I wrote the latter one ;D Was trying to copy it here, but that is not so simple. Have to think a bit more.

**Comment by mrazauskas** [comment]
@Maxim-Mazurok Could I ask you to take a look at this issue?

You implement index signatures support for `spyOn` in #60971. Something goes wrong in the case describe above. Unfortunately I can't fully understand some details in your approach.

The implementation in Jest repo is different. All works fine with this case is types are imported from `@jest/globals`. Strangely if I copy that implementation here, some things get broken. Might be there is some detail which I cannot grasp.

**Comment by Maxim-Mazurok** [comment, marked as answer]
**TL;DR**: potential fix: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63182

Sure, it's quite a tedious process of breaking down all the types and utilities and checking with VS Code and/or [ts-type-expand](https://marketplace.visualstudio.com/items?itemName=kimuson.ts-type-expand) extension.

To summarise, when doing `jest.spyOn(Storage.prototype, 'setItem')` with `@types/jest@29.2.2`,

my findings from `ConstructorPropertyNames`:
- `keyof T` will be `string | number`, because of index signature `[name: string]: any;` I think; same problem as I had in https://github.com/facebook/jest/issues/12966 IIRC
- now `T[K]` will be `any`
- and now `T[K] extends Constructor` = `any extends Constructor` = true, so the `spyOn` overload with `ConstructorPropertyNames` will be matched which is not correct, it won't be matched for `Event` or `Date`, only for `Storage` because of the index signature

my findings from `spyOn` overload with `ConstructorPropertyNames`:
- the `ConstructorPropertyNames<Required<T>>` returns `string` (while it's probably expected to return `never`, same as for `Event` of `Date`)
- `Required<T>[M]` will be `Required<typeof Storage.prototype>['setItem']` = `(key: string, value: string) => void`
- And `(key: string, value: string) => void` does not extent `Constructor` (aka `new (...args: any[])`) so `never` will be returned.

Here is some code I used for debugging:
```ts
type Constructor = new (...args: any[]) => any;
type ConstructorPropertyNames<T> = { [K in keyof T]: T[K] extends Constructor ? K : never }[keyof T] & string;

type spyOnType<T extends {}, M extends ConstructorPropertyNames<Required<T>>> = (
	object: T,
	method: M,
) => Required<T>[M] extends new (...args: any[]) => any ? 'bla' : never;

type tT = typeof Storage.prototype;
type tM = 'setItem'; // ConstructorPropertyNames<Required<tT>>;
type reqTmp = Required<tT>[tM];
type retTmp = reqTmp extends new (...args: any[]) => any ? 'yes' : never;

type tmp = ReturnType<spyOnType<typeof Storage.prototype, 'setItem'>>;

type kindaT = Required<typeof Storage.prototype>;

// string | number
type keys = keyof kindaT;

// any
type kindaTK1 = kindaT[string];
// any
type kindaTK2 = kindaT[number];

// "of course"
type extendsConstructorDefinitely = (new () => void) extends Constructor ? 'of course' : 'no way';

// "of course" | "no way"                           // kindaTK1
type extendsConstructorAndDoesNotExtendAtTheSameTime = any extends Constructor ? 'of course' : 'no way';

type temp = { [K in keyof kindaT]: kindaT[K] extends Constructor ? K : never };
type temp2 = temp[keys] & string;
// type temp3 = temp[number];

const x = jest.spyOn(Storage.prototype, 'setItem');
type T_fromSpyOn = kindaT;
type M_fromSpyOn = ConstructorPropertyNames<kindaT>;

type y = Required<T_fromSpyOn>[M_fromSpyOn] extends new (...args: any[]) => any ? 'yes' : 'no';

jest.spyOn(Storage.prototype, 'setItem').mockImplementation(jest.fn());
```

~~I don't really know how to fix this tho.~~
- `keyof typeof Storage` = `"prototype"` (no new/constructor)
- `keyof typeof Storage.prototype` = `string | number` (no new/constructor)
- Even if we remove the index signature, we'll get `keyof typeof Storage.prototype` = `"clear" | "getItem" | "key" | "removeItem" | "setItem"` but no new/constructor in that union. This means that `[K in keyof T]: T[K] extends Constructor` doesn't make sense as it will never find a constructor key.

I see, we don't spy on the constructor using new/constructor index, we just wrap the class in an object, like so: `const myModule = { Storage }`, etc.
Here's PR with a fix, please feel free to review it: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63182

**Reply by mrazauskas** [reply]
Thanks for quick response and all the details. Very interesting! I did not know about the `ts-type-expand` extension. Useful one to have around.

**Reply by Maxim-Mazurok** [reply]
No worries! Hopefully that will be my first accepted discussion answer 😅

**Reply by Maxim-Mazurok** [reply]
@wimbarelds the fix had been merged just now, please give it a try once it's published to npm and consider marking this answer as accepted if it works, cheers!

**Reply by wimbarelds** [reply]
My apologies, at my (e-commerce) job npm updates have been frozen around black friday, but will have a look early next week.

**Reply by Maxim-Mazurok** [reply]
No worries :)
